File: LeetCode/Hao/274HIndex.java
public class Solution {
    public int hIndex(int[] citations) {
        
        int[] freq = new int[citations.length + 1];
        
        for (int i = 0; i < citations.length; i ++) {
            if (citations[i] > citations.length) {
                freq[citations.length] ++;
            } else {
                freq[citations[i]] ++;
            }
        }
        int sum = 0;
        int i;
        
        for (i = citations.length; i >= 0; i --) {

            sum += freq[i];
            if (sum >= i){
                break;
            }
        }
        
        return i;
    }
}

File: LeetCode/Peng/274H-Index.cpp
//After we sorted the vector,we will know that all the paper that after //i will have the same or more citations.
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        int result = 0;
        sort(citations.begin(), citations.end());
        for (int i = 0; i < n; i++) {
            result = max(result, min(n - i, citations.at(i)) );
        }
        return result;
    }
};


class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        int result = 0;
        vector<int> count(n + 2, 0);
        for (int i = 0; i < n; i++) {
            if (citations.at(i) > n) {
                count.at(n)++;
            } else {
                count.at(citations.at(i))++;
            }
        }
        for (int i = n; i >= 0; i--) {
            count.at(i) += count.at(i + 1);
            if (count.at(i) >= i) {
                result = i;
                break;
            }
        }
        return result;
    }
};